### PR TITLE
Translate EMS Example Files to PythonPlugin

### PR DESCRIPTION
**Pull request by mitchute**
Pull request overview
---------------------

Translate the remaining EMS example files to PythonPlugin. Some current python example files don't mirror what the original EMS file does, so this should be checked and results between the two methods compared.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by Myoldmopar** [comment]
And....assuming CI is OK with this last commit, we are done.  This one was indeed a beast, but it seems to be working well.  Spot checking a few zone temps reveals good agreement but then also things smoothing out like this:

image

There are small warmup convergence diffs, and some other minor stuff, but overall it seems to be in great shape.

**Comment by mitchute** [comment]
Agreed, @Myoldmopar. This is ready to go. I checked the diffs on this last file (DemandManager_LargeOffice) and it's the worst offender regarding math differences. Some smoothing in places but overall it looks good. 

Var_0
Var_1
Var_2
Var_4
Var_5

File: src/EnergyPlus/EMSManager.cc [hunk 1]
```diff
@@ -304,7 +304,10 @@ namespace EMSManager {
         anyProgramRan = false;
         if (!AnyEnergyManagementSystemInModel) return; // quick return if nothing to do
 
-        if (iCalledFrom == DataGlobals::emsCallFromBeginNewEvironment) BeginEnvrnInitializeRuntimeLanguage();
+        if (iCalledFrom == DataGlobals::emsCallFromBeginNewEvironment) {
+            BeginEnvrnInitializeRuntimeLanguage();
+            PluginManagement::onBeginEnvironment();
+        }
```
Comment (Myoldmopar) [inline src/EnergyPlus/EMSManager.cc:310]: Added a new begin environment calling point in the plugins to clear trends and reset stuff.  This is a big improvement.

File: src/EnergyPlus/PluginManager.cc [hunk 2]
```diff
@@ -422,8 +433,8 @@ namespace PluginManagement {
 #if LINK_WITH_PYTHON == 1
         for (auto &plugin : plugins) {
             if (plugin.runDuringWarmup || !DataGlobals::WarmupFlag) {
-                plugin.run(iCalledFrom);
-                anyRan = true;
+                bool const didOneRun = plugin.run(iCalledFrom);
+                if (didOneRun) anyRan = true;
```
Comment (Myoldmopar) [inline src/EnergyPlus/PluginManager.cc:437]: Properly return whether a plugin *actually* ran, not just if we tried to run plugins.  This is another big improvement.

File: testfiles/CMakeLists.txt [hunk 3]
```diff
@@ -415,14 +415,23 @@ ADD_SIMULATION_TEST(IDF_FILE PurchAirWithDaylightingAndShadeControl.idf EPW_FILE
...
+ADD_SIMULATION_TEST(IDF_FILE PythonPluginPlantOperation_largeOff.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
+ADD_SIMULATION_TEST(IDF_FILE PythonPluginReplaceTraditionalManagers_LargeOffice.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 ADD_SIMULATION_TEST(IDF_FILE PythonPluginTestMathAndKill.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw EXPECT_FATAL)  # Expected to Fatal
+ADD_SIMULATION_TEST(IDF_FILE PythonPluginThermochromicWindow.idf EPW_FILE USA_NV_Las.Vegas-McCarran.Intl.AP.723860_TMY3.epw)
 ADD_SIMULATION_TEST(IDF_FILE PythonPluginUserDefined5ZoneAirCooled.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
+ADD_SIMULATION_TEST(IDF_FILE PythonPluginUserDefinedWindACAuto.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
+ADD_SIMULATION_TEST(IDF_FILE PythonPluginWindowShadeControl.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
+ADD_SIMULATION_TEST(IDF_FILE PythonPluginLrgOff_GridStorageSmoothing.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
+ADD_SIMULATION_TEST(IDF_FILE PythonPluginCustomTrendVariable.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
```
Comment (Myoldmopar) [inline testfiles/CMakeLists.txt:434]: Added gobs of new python plugin test cases that match the EMS cases.  We are showing we can match the results of EMS which is a huge step.

File: testfiles/EMSCurveOverride_PackagedTerminalHeatPump.idf [hunk 4]
```diff
@@ -2115,3 +2115,6 @@
     HTML;                    !- Column Separator
 
 !-   ===========  ALL OBJECTS IN CLASS: OUTPUT:VARIABLE ===========
+  Output:Variable,Zone1PTHPDXCoolCoil,Cooling Coil Total Cooling Rate,timestep;
+  Output:Variable,Zone2PTHPDXCoolCoil,Cooling Coil Total Cooling Rate,timestep;
+  Output:Variable,Zone3PTHPDXCoolCoil,Cooling Coil Total Cooling Rate,timestep;
```
Comment (Myoldmopar) [inline testfiles/EMSCurveOverride_PackagedTerminalHeatPump.idf]: Just added a few extra outputs to a couple EMS files for debugging, they aren't hurting anything but I may take them out if I need to do more commits.

File: testfiles/PythonPluginConstantVolumePurchasedAir.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+            # get handles if needed
+            if self.need_to_get_handles:
+                # self.Shade_Status_None_handle = self.api.exchange.get_global_handle(
+                #     "Shade_Status_None")
+
+                self.Shade_Status_Off_handle = self.api.exchange.get_global_handle(
+                    "Shade_Status_Off")
+
+                # self.Shade_Status_Interior_Shade_On_handle = self.api.exchange.get_global_handle(
```
Comment (Myoldmopar) [inline testfiles/PythonPluginConstantVolumePurchasedAir.py]: All these commented pieces should be cleaned out.

**Review by Myoldmopar — COMMENTED** [review]
I'm going to make a commit with two things: some commented Python cleaned out, and the extra output variables added to the two EMS files cleaned out.  It will not need to wait on CI for new results.

**Comment by Myoldmopar** [comment]
Thanks @mitchute and @dareumnam, this is a major step for the python work.

**Comment by rraustad** [comment]
I'm not sure but the plot of Core Top: Zone Mean Air Temperature makes me think of reduced system time steps and whether python is picking up information correctly during this time. A detailed simulation may show issues?

**Pull request merged**